cloudinit/performance.py

```diff
@@ -0,0 +1,101 @@
+import functools
+import logging
+import time
+
+LOG = logging.getLogger(__name__)
+
+
+class Timed:
+    """
+    A context manager which measures and optionally logs context run time.
+
+    :param msg: A message that describes the thing that is being measured
+    :param threshold: Threshold, in seconds. When the context exceeds this
+        threshold, a log will be made.
+    :param log_mode: Control whether to log. Defaults to "threshold". Possible
+        values include:
+        "always" - Always log 'msg', even when 'threshold' is not reached.
+        "threshold" - Log when context time exceeds 'threshold'.
+        "skip" - Do not log. Context time and message are stored in the
+            'output' and 'delta' attributes, respectively. Used to manually
+            coalesce with other logs at the call site.
+
+    usage:
+
+        this call:
+        ```
+        with Timed("Configuring the network"):
+            run_configure()
+        ```
+
+        might produce this log:
+        ```
+            Configuring the network took 0.100 seconds
+        ```
+    """
+
+    def __init__(
+        self,
+        msg: str,
+        /,
+        *,
+        threshold: float = 0.01,
+        log_mode: str = "threshold",
```

Comment:
I don't love this parameter being a magic string, especially when it only takes two values. Is there a reason setting threshold to 0 isn't a valid way to signal to log always? If so, could this field be a boolean like `always_log` or something similar?

Reply:
There are technically three (see "skip" in the docstring), not two, but the behavior of the third is implicit, which should probably be made explicit for code clarity.

> Is there a reason setting threshold to 0 isn't a valid way to signal to log always?

It would work, but then you may end up with messages ending with `.. took 0.000 seconds to complete`, or other numbers which come in under the threshold value, which is useless. By having three options `"skip", "always", "threshold"`, we're able to maximize signal to noise ratio.

> I don't love this parameter being a magic string

Nor do I, but it seems like the Pythonic way to control the mode of a function which has more than two modes. Consider the arguments in `"".encode()`'s, and `b''.decode()`. An alternative that I considered briefly was to define enums, but that requires more names to be imported and I wasn't able to find any examples of this in the Python standard library so this approach seems preferred.

@TheRealFalcon Thoughts? If you don't mind keeping it as a string, I can make the `"skip"` logic explicit.

Reply:
Ah, thanks. I didn't realize the "skip" option. I go back and forth about enums, but the string option here makes the most sense.